Stripe/STPSourceInfoViewController.m

```diff
@@ -0,0 +1,307 @@
+#import "STPSource+Private.h"
+#import "STPTextFieldTableViewCell.h"
+
+@interface STPSourceInfoViewController () <UITableViewDelegate, UITableViewDataSource, STPTextFieldTableViewCellDelegate>
+
+@property(nonatomic)UIBarButtonItem *doneItem;
+@property(nonatomic)STPSourceParams *sourceParams;
+@property(nonatomic)NSArray<STPTextFieldTableViewCell *>*cells;
+@property(nonatomic)STPSourceType sourceType;
+
+@end
+
+@implementation STPSourceInfoViewController
+
++ (BOOL)canCollectInfoForSourceType:(STPSourceType)type {
+    switch (type) {
+        case STPSourceTypeBancontact:
+        case STPSourceTypeGiropay:
+        case STPSourceTypeIDEAL:
+        case STPSourceTypeSofort:
+            return YES;
+        default:
+            return NO;
+    }
+}
+
+- (instancetype)initWithSourceParams:(STPSourceParams *)sourceParams
+                               theme:(STPTheme *)theme {
+    self = [super initWithTheme:theme];
+    STPSourceType sourceType = [STPSource typeFromString:sourceParams.type];
+    if (![[self class] canCollectInfoForSourceType:sourceType]) {
+        return nil;
+    }
+    if (self) {
+        _sourceParams = sourceParams;
+        _sourceType = sourceType;
+        self.title = [self titleForSourceType:sourceType];
+    }
+    return self;
+}
+
+- (NSString *)titleForSourceType:(STPSourceType)type {
+    switch (type) {
+        case STPSourceTypeBancontact:
+            return STPLocalizedString(@"Bancontact Info", @"Title for form to collect Bancontact account info");
+        case STPSourceTypeGiropay:
+            return STPLocalizedString(@"Giropay Info", @"Title for form to collect Giropay account info");
+        case STPSourceTypeIDEAL:
+            return STPLocalizedString(@"iDEAL Info", @"Title for form to collect iDEAL account info");
+        case STPSourceTypeSofort:
+            return STPLocalizedString(@"Sofort Info", @"Title for form to collect Sofort account info");
+        default:
+            return @"";
+    }
+}
+
+- (void)createAndSetupViews {
+    [super createAndSetupViews];
+    UIBarButtonItem *doneItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(nextPressed:)];
+    self.doneItem = doneItem;
+    self.stp_navigationItemProxy.rightBarButtonItem = doneItem;
+    self.stp_navigationItemProxy.rightBarButtonItem.enabled = NO;
+
+    self.tableView.dataSource = self;
+    self.tableView.delegate = self;
+
+    [self createAndSetupCells];
+
+    [self.view addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(endEditing)]];
+}
+
+- (void)createAndSetupCells {
+    STPTextFieldTableViewCell *nameCell = [[STPTextFieldTableViewCell alloc] init];
+    nameCell.placeholder = STPLocalizedString(@"Name", @"Caption for Name field on bank info form");
+    if (self.sourceParams.owner) {
+        nameCell.contents = self.sourceParams.owner[@"name"];
+    }
+    switch (self.sourceType) {
```

Comment:
Should we try to pull out the source-specific logic from this class into some other delegate or data source?